### DISCUSSION TITLE
Mark "case" keyword as optional in pattern "match"

### DISCUSSION THREAD

**Discussion by jaisw7** [scala/scala3#19599] [Feature Requests]
Instead of 
```
val result = x match {
  case 0 => "Zero"
  case n if n > 0 => "Positive"
  case n if n < 0 => "Negative"
  case _ => "Unknown"
}
```

I believe 
```
val result = x match {
  0 => "Zero"
  n if n > 0 => "Positive"
  n if n < 0 => "Negative"
  _ => "Unknown"
}
```
is more concise. 

I was wondering if there is a particular reason for using `case` keyword e.g., it makes the grammar unambiguous? If not, can we keep it as optional? I can submit a patch if everyone is aligned.

**Comment by som-snytt** [comment]
The RHS is a block; `case` tells you where it ends without braces. I proposed omitting the wildcard, `case => "Unknown"`, that was already too much (or too little). There was a pre-sip for `x match p` to mean `case p => true case _ => false`. I think there is an open discussion about improving guard syntax (for disjoint cases).

**Comment by jaisw7** [comment]
Got it. I realized that OCaml and Rust uses symbols for disambiguation:

```ocaml
let result = match x with 
|  0 -> "Zero"
|  n when n > 0 -> "Positive"
|  n when n < 0 -> "Negative"
|  _ -> "Unknown"
```

```rust
let result = match x {
    0 => "Zero"
  , n if n > 0 => "Positive"
  , n if n < 0 => "Negative"
  , _ => "Unknown"
}
```

I wish Scala was as minimal as Ocaml and/or Rust.

**Comment by spamegg1** [comment]
I also felt as @jaisw7 initially about the `case` keyword, but my view has changed over time. Sometimes things can be a bit too concise. Having a meaningful, readable (and still short) guard is nice. So many easy-to-make mistakes with forgetting SML / OCaml's `|`, not to mention dealing with the weird alignment / indentation.

As another, and arguably more important, defense of `case`, it creates conceptual consistency, which is especially useful in teaching. `case` being used in pattern matching, `enum`s and `case class`es make things very clear and easy to understand. Moreover it perfectly matches (no pun intended) mathematical proof-writing where the word *case* is used, which makes the code match (again) the proof one-to-one.
